### PR TITLE
feat(header): update collection header styles COMPASS-5519

### PR DESCRIPTION
**Pull request by alenakhineika**
## Description

Update styles and components used in the collection header. This is the area with the namespace and collection stats. To simplify the codebase and get rid of the reflux store, combine the collection stats and collection packages into one new package.

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [x] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-components/src/index.ts
Comment (gribnoysup) [inline packages/compass-components/src/index.ts:38]: This one is really low level, so I wanted to ask if this is something we are planning to do more now, using Box instead of just html elements?

Comment (alenakhineika) [reply packages/compass-components/src/index.ts:38]: Not sure. They kind of use it as a link here in example: https://www.mongodb.design/component/box/example/
When a link from typography has a very specific style with an icon: https://www.mongodb.design/component/typography/example/
I am ok with having just `<a></a>` here, but I thought the idea was to use leafy green as much as possible.

Comment (gribnoysup) [reply packages/compass-components/src/index.ts:38]: Actionable elements on the screen should be buttons or links otherwise they can't be accessed by keyboard. If you need it to look like a link, you can check out how we do this in other places:

https://github.com/mongodb-js/compass/blob/b06c5ed05524d60d4705c7bea0df33b018a7342a/packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx#L55-L66

File: packages/compass-collection/src/modules/stats.ts
Comment (Anemy) [inline packages/compass-collection/src/modules/stats.ts:24]: Small idea or suggestion, since this `StatsObject` looks like it always pertains to the stats of a collection, how do you feel about the name `CollectionStats`?
```suggestion
export interface CollectionStats {
```

File: packages/compass-collection/src/components/collection/collection.tsx
Comment (Anemy) [inline packages/compass-collection/src/components/collection/collection.tsx:110]: Giving `useEffect` no second argument acts as both componentDidMount and componentDidUpdate, as in it runs first on mount and then on every re-render. Do we want to keep this using the `[]`? I'm not sure I follow the change. If we want to track the sub tab changes maybe we make it use `activeSubTab` and `tabs` as dependencies?

**Review by Anemy — COMMENTED** [review]
Nice work, I like that we're doing some simplification here. I left a couple comments, mostly on some ways we might improve types now that we have types here

File: packages/compass-collection/src/components/collection/collection.tsx
Comment (alenakhineika) [reply packages/compass-collection/src/components/collection/collection.tsx:110]: Lint complaints with []: `React Hook useEffect has missing dependencies: 'activeSubTab' and 'tabs'. Either include them or remove the dependency array.`

What would be the proper approach?

Comment (Anemy) [reply packages/compass-collection/src/components/collection/collection.tsx:110]: Lint is showing that error because function is using the `activeSubTab` and `tabs` props inside of it. Because of the function's scope those variables could be stale in the useEffect, or not the current, correct value if we pass only `, []` because the function is only set on first mount. This might be alright if we only want it to run once, but in this case do we want the tracking to occur every time the `activeSubTab` changes? 
If we want the tracking event every time the `activeSubTab` or `tabs` change we can add them to the dependency list like this:
```suggestion
  }, ['activeSubTab', 'tabs']);
```

More information on how `useEffect` and other hooks use these dependencies and why the `exhaustive-deps` lint shows up can be found here: https://reactjs.org/docs/hooks-reference.html#conditionally-firing-an-effect

Comment (Anemy) [reply packages/compass-collection/src/components/collection/collection.tsx:110]: If we want the event to only occur when the `activeSubTab` changes and not the `tabs` we could use another hook, `useMemo` to track the active subtab name and then add that new variable to this dependency list.
Something like:

```ts
const activeSubtabName = useMemo(() => trackingIdForTabName((tabs && tabs[activeSubTab]) || 'Unknown'), ['activeSubTab', 'tabs']);
```

```suggestion
  }, [activeSubtabName]);
```

File: packages/compass-components/src/index.ts
Comment (alenakhineika) [reply packages/compass-components/src/index.ts:38]: Thx for sharing the example! Will do this change!

File: packages/compass-collection/src/modules/stats.ts
Comment (gribnoysup) [reply packages/compass-collection/src/modules/stats.ts:76]: We do have collection-model typedefs

File: packages/compass-collection/src/modules/stats.ts
Comment (alenakhineika) [reply packages/compass-collection/src/modules/stats.ts:24]: It will overlap with the component name then, but I will make it `CollectionStatsObject` to be more explicit.

File: packages/compass-collection/src/components/collection-header-actions/view-information.tsx
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-header-actions/view-information.tsx:7]: Is it allowed to use spacing for font sizes and line heights? I can't see this font size in leafygreen anywhere

File: packages/compass-collection/src/components/collection-header/collection-header.tsx
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-header/collection-header.tsx:70]: Seems like we are duplicating quite a lot of styles between light and dark here, would it make sense to have only properties that change defined separately?

File: packages/compass-collection/src/components/collection-stats-item/collection-stats-item.tsx
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-stats-item/collection-stats-item.tsx:33]: Is it okay that we use these very custom font sizes here? My understanding was that we stick to what typography packages allows us to use

File: packages/compass-collection/src/components/collection/collection.tsx
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection/collection.tsx:29]: Same as above, wouldn't it make sense to have only styles that change extracted? This should make maintaining this a bit easier over time

File: packages/compass-collection/src/stores/index.ts
Comment (gribnoysup) [inline packages/compass-collection/src/stores/index.ts:174]: This `metadata` object doesn't have `stats` available

File: packages/compass-collection/src/stores/index.ts
Comment (gribnoysup) [inline packages/compass-collection/src/stores/index.ts:91]: I see that you have `stats` as a slice now and it inherits most of the previous plugin logic (like getting updated when collection status updates), but it doesn't look like it's ever used in the collections as you are also store stats info with every tab and this is what gets passed to the header. I think we should decide which approach to take here: either have the singleton logic copied over, or keep stats data copied over to all the tabs, but then keeping them in sync when collection status changes (this is not happening at the moment, you can reproduce by opening a collection from the sidebar without opening collections list first)

File: packages/compass-collection/src/components/collection-header-actions/view-information.tsx
Comment (alenakhineika) [reply packages/compass-collection/src/components/collection-header-actions/view-information.tsx:6]: I don't know, I have seen it in other places in compass implemented like this, so I thought you have some kind of agreement. What should be used instead?

File: packages/compass-collection/src/components/collection-header/collection-header.tsx
Comment (alenakhineika) [reply packages/compass-collection/src/components/collection-header/collection-header.tsx]: I was also considering this but was wondering if it reduces code readability with lots of styles combined with each other.

File: packages/compass-collection/src/components/collection-stats-item/collection-stats-item.tsx
Comment (alenakhineika) [reply packages/compass-collection/src/components/collection-stats-item/collection-stats-item.tsx:33]: With sizes out of the box, it doesn't match the design 😞

File: packages/compass-collection/src/components/collection-header/collection-header.tsx
Comment (Anemy) [reply packages/compass-collection/src/components/collection-header/collection-header.tsx]: While it does introduce more complexity at the rendering level I think I would agree that separating the theme styles from the main component styles that both share is probably worth doing, or it's at least my preference. I think having general css rules like `flexShrink` or `overflow` in both of the styles can fairly easily lead to one style being out of sync with the other if one is changed without the other.

Here's how I was doing it in `find-in-page`:
https://github.com/mongodb-js/compass/blob/3bfdea024eff1067e282681d729039fb1c6301b4/packages/compass-find-in-page/src/components/find-in-page-input.tsx#L138
It's definitely more things going on in the rendering, but I think `cx` might help improve readability concerns.

File: packages/compass-collection/src/components/collection/collection.tsx
Comment (alenakhineika) [reply packages/compass-collection/src/components/collection/collection.tsx]: Returned []!

File: packages/compass-collection/src/components/collection-stats-item/collection-stats-item.tsx
Comment (gribnoysup) [reply packages/compass-collection/src/components/collection-stats-item/collection-stats-item.tsx:33]: Hmmm, shouldn't designs be using spacings from leafygreen though? @Anemy is it some special case here that we decided doesn't need to be aligned with leafygreen?

File: packages/compass-collection/src/components/collection-header-actions/view-information.tsx
Comment (alenakhineika) [reply packages/compass-collection/src/components/collection-header-actions/view-information.tsx]: I replaced all places in `compass-collection` where I used spacing for font sizes with some elements from typography. Pls let me know if this is ok I need to update it in some other way. I also removed line heights.

File: packages/compass-collection/src/stores/index.ts
Comment (alenakhineika) [reply packages/compass-collection/src/stores/index.ts]: Fixes opening new tabs with updated stats info. Hope I understood the data flow correctly. Pls, let me know if misused something.

**Comment by gribnoysup** [comment]
Something in this PR is causing a very weird e2e test failure, I'm trying to figure out what it is

File: packages/compass-collection/src/components/collection-stats-item/collection-stats-item.tsx
Comment (Anemy) [reply packages/compass-collection/src/components/collection-stats-item/collection-stats-item.tsx]: Sorry for missing this, looks like it's gone now so I'll resolve. I think we typically should avoid using `lineHeight` if possible. We should aim to use `spacings` from leafygreen as much as possible. There are rare cases where I think it makes sense to deviate and use custom styles though.

File: packages/compass-collection/src/components/collection-header/collection-header.tsx
Comment (Anemy) [inline packages/compass-collection/src/components/collection-header/collection-header.tsx:29]: Not totally related, looks like this background might not be needed anymore? If we remove this background setting it looks like the background adjusts based on the background color. (with the env variable `COMPASS_LG_DARKMODE`)
```suggestion
```

Screen Shot 2022-04-20 at 5 59 13 PM
Screen Shot 2022-04-20 at 6 03 11 PM

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.tsx
Comment (Anemy) [inline packages/compass-collection/src/components/collection-header-actions/collection-header-actions.tsx:12]: Maybe
```suggestion
    flexShrink: 0,
    marginLeft: spacing[2],
```

Screen Shot 2022-04-20 at 6 07 05 PM
Screen Shot 2022-04-20 at 6 08 32 PM

File: packages/compass-collection/src/modules/stats.ts
Comment (Anemy) [inline packages/compass-collection/src/modules/stats.ts:91]: Should we 
```suggestion
  let stats = { ...INITIAL_STATE };
```
here to ensure we don't mutate the `INITIAL_STATE` object?

File: packages/compass-collection/src/modules/stats.ts
Comment (Anemy) [inline packages/compass-collection/src/modules/stats.ts:125]: Same as above, would object spreading this make the INTIAL_STATE and `action.stats;` usage a little safer?

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.tsx
Comment (gribnoysup) [reply packages/compass-collection/src/components/collection-header-actions/collection-header-actions.tsx:12]: Sometimes I feel like doing a single PR that just puts `flex: none` on literally all the buttons in the app, I don't think we ever want them to collapse like that 😅

**Comment by gribnoysup** [comment]
@Anemy following up on our discussion [here's a fix](https://github.com/mongodb-js/compass/pull/2965/commits/1663bf7972633f98c8a0c7483c0c096674707e76) I came up with: we are still cleaning up the tabs on disconnect, and at the same time the `all-tabs-closed` is now only emitted when all tabs are closed by the user and not programmatically, check it out and tell me if that makes sense to you

File: packages/compass-collection/src/components/collection-header/collection-header.tsx
Comment (Anemy) [inline packages/compass-collection/src/components/collection-header/collection-header.tsx:26]: Do we need a min height here? I'm thinking letting the elements define the height and adding some bottom padding here would be a solution the might make the height not required to change if the elements shrink in new leafygreen iterations for instance?

**Review by Anemy — APPROVED** [review]
lgtm! A couple open questions on a few places we might be able to do some improvements probably worth looking into before merging.

I like that the e2e failure fix you added removes the very general `store.subscribe` which definitely could have led to similar issues in future, nice.

File: packages/compass-collection/src/components/collection-header/collection-header.tsx
Comment (Anemy) [reply packages/compass-collection/src/components/collection-header/collection-header.tsx:26]: We need it for keeping consistent height when swapping between views and non-views as the elements on the right are a different height.

**Comment by Anemy** [comment]
Added a commit updating some styles and fixing some overflow with long collection names.

**Comment by gribnoysup** [comment]
Looks good to me. Do we have tickets for fixing all the any types that we are introducing here? Might be worth adding so we don't forget to clean them up

File: packages/compass-collection/src/components/collection-header/collection-header.tsx
Comment (alenakhineika) [inline packages/compass-collection/src/components/collection-header/collection-header.tsx:26]: To be honest this looks very weird :) Don't we know for sure that spacing value is a number? We could add an extra step to `compass-components` to ensure that it is a number or 0 if someone misused an index.

Comment (gribnoysup) [reply packages/compass-collection/src/components/collection-header/collection-header.tsx:26]: Ah, that's on me, for some reason TS doesn't compile for me locally if I don't explicitly convert these to numbers, if it doesn't cause errors for y'all totally feel free to remove this

Comment (alenakhineika) [reply packages/compass-collection/src/components/collection-header/collection-header.tsx:26]: Ah got ya! You probably have a newer version than we do. From my experience, it is better to keep it then, because later we will all encounter it anyway :)

**Pull request merged**
